// @testquality/sdk/dist/gen/domain/run/runService.js
"use strict";
/**
 * Copyright (C) 2021 BitModern, Inc - All Rights Reserved
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.runCreateMany = exports.runCreateOne = exports.runUpdateMany = exports.runUpdateOne = exports.runDeleteMany = exports.runDeleteOne = exports.runGetOne = exports.runGetMany = void 0;
const ClientSdk_1 = require("../../../ClientSdk");
const getResponse_1 = require("../../actions/getResponse");
const chunkArray_1 = require("../../actions/chunkArray");
const Routes_1 = require("../../routes/Routes");
const runGetMany = (queryParams) => {
    var _a, _b;
    const config = {
        method: 'get',
        url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)(),
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.runGetMany = runGetMany;
const runGetOne = (id, queryParams) => {
    var _a, _b;
    const config = {
        method: 'get',
        url: `${(_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)()}/${id}`,
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.runGetOne = runGetOne;
const runDeleteOne = (id, queryParams) => {
    var _a, _b;
    const config = {
        method: 'delete',
        url: `${(_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)()}/${id}`,
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.runDeleteOne = runDeleteOne;
const runDeleteMany = (data, queryParams) => {
    const chunks = (0, chunkArray_1.chunkArray)(data, 1000);
    return Promise.all(chunks.map((chunk) => {
        var _a, _b;
        const config = {
            method: 'post',
            url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)() + '/delete',
            params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
            list: chunk,
            headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
        };
        return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
            ? queryParams.batch.addBatch(config)
            : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
    }));
};
exports.runDeleteMany = runDeleteMany;
const runUpdateOne = (id, data, queryParams) => {
    var _a, _b;
    const config = {
        method: 'put',
        url: `${(_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)()}/${id}`,
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        data,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.runUpdateOne = runUpdateOne;
const runUpdateMany = (data, queryParams) => {
    const chunks = (0, chunkArray_1.chunkArray)(data, 1000);
    return Promise.all(chunks.map((chunk) => {
        var _a, _b;
        const config = {
            method: 'post',
            url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)(),
            params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
            list: chunk,
            headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
        };
        return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
            ? queryParams.batch.addBatch(config)
            : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
    }));
};
exports.runUpdateMany = runUpdateMany;
const runCreateOne = (data, queryParams) => {
    var _a, _b;
    const config = {
        method: 'post',
        url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)(),
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        data,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.runCreateOne = runCreateOne;
const runCreateMany = (data, queryParams) => {
    const chunks = (0, chunkArray_1.chunkArray)(data, 1000);
    return Promise.all(chunks.map((chunk) => {
        var _a, _b;
        const config = {
            method: 'post',
            url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : (0, Routes_1.RunRoute)(),
            params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
            list: chunk,
            headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
        };
        return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
            ? queryParams.batch.addBatch(config)
            : (0, getResponse_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
    }));
};
exports.runCreateMany = runCreateMany;
