// @testquality/sdk/dist/services/testRail/api/projectsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postImportProject = exports.getProjects = void 0;
const ClientSdk_1 = require("../../../ClientSdk");
const actions_1 = require("../../../gen/actions");
const getProjects = (queryParams) => {
    var _a;
    const config = {
        method: 'get',
        url: '/testrail/projects',
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, actions_1.getResponse)((_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _a !== void 0 ? _a : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.getProjects = getProjects;
const postImportProject = (testRailProjectId, tqProjectId, mapping, queryParams) => {
    var _a;
    const config = {
        method: 'post',
        url: '/testrail/projects/import',
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        data: {
            projectId: testRailProjectId,
            tqProjectId,
            entitiesMapping: mapping,
        },
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, actions_1.getResponse)((_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _a !== void 0 ? _a : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.postImportProject = postImportProject;
