// @testquality/sdk/dist/gen/routes/Routes.js
"use strict";
/**
 * Copyright (C) 2021 BitModern, Inc - All Rights Reserved
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientRoute = exports.CheckSuiteCheckSuitePullRequestRoute = exports.CheckSuiteCheckRunRoute = exports.CheckSuiteRoute = exports.CheckRunRunRoute = exports.CheckRunRoute = exports.CheckListItemCheckListItemUserRoute = exports.CheckListItemRoute = exports.CheckListCheckListItemRoute = exports.CheckListRoute = exports.CaseTypeTestRoute = exports.CaseTypeLabelAssignedRoute = exports.CaseTypeCaseTypeProjectRoute = exports.CaseTypeRoute = exports.CasePriorityTestRoute = exports.CasePriorityLabelAssignedRoute = exports.CasePriorityCasePriorityProjectRoute = exports.CasePriorityRoute = exports.CapabilityProjectCapabilityDefaultRoute = exports.CapabilityIntegrationRoute = exports.CapabilityRoute = exports.BillingContactRoute = exports.BaseIntegrationIntegrationRoute = exports.BaseIntegrationBaseCapabilityRoute = exports.BaseIntegrationAppConfigRoute = exports.BaseIntegrationRoute = exports.BaseCapabilityCapabilityRoute = exports.BaseCapabilityBaseIntegrationRoute = exports.BaseCapabilityRoute = exports.AttachmentRoute = exports.AppUserPullRequestRoute = exports.AppUserBranchRoute = exports.AppUserAppInstallRoute = exports.AppUserRoute = exports.AppInstallProjectPullRequestRoute = exports.AppInstallProjectIntegrationProjectRoute = exports.AppInstallProjectCheckSuiteRoute = exports.AppInstallProjectCheckRunRoute = exports.AppInstallProjectBranchRoute = exports.AppInstallProjectRoute = exports.AppInstallAppInstallProjectRoute = exports.AppInstallAppInstallIntegrationRoute = exports.AppInstallRoute = exports.AppConfigAppInstallRoute = exports.AppConfigRoute = exports.AccessRoleUserRoute = exports.AccessRoleShareRoute = exports.AccessRoleProjectRoute = exports.AccessRolePolicyRoute = exports.AccessRoleRoute = void 0;
exports.ExplorationRoute = exports.EnvironmentRunResultRoute = exports.EnvironmentExplorationItemRoute = exports.EnvironmentEnvironmentResourceRoute = exports.EnvironmentEnvironmentProjectRoute = exports.EnvironmentEnvironmentPlanRoute = exports.EnvironmentEnvironmentExplorationRoute = exports.EnvironmentRoute = exports.DocTypeDocTemplateRoute = exports.DocTypeDocRoute = exports.DocTypeComponentDocTypeRoute = exports.DocTypeRoute = exports.DocTemplateDocRoute = exports.DocTemplateComponentDocTemplateRoute = exports.DocTemplateRoute = exports.DocComponentDocRoute = exports.DocAttachmentRoute = exports.DocRoute = exports.DefectUserRoute = exports.DefectTypeRoute = exports.DefectStatusRequirementRoute = exports.DefectStatusNativeDefectStatusRoute = exports.DefectStatusDefectRoute = exports.DefectStatusRoute = exports.DefectRunResultRoute = exports.DefectResRequirementRoute = exports.DefectResNativeDefectResRoute = exports.DefectResDefectRoute = exports.DefectResRoute = exports.DefectPriorityRoute = exports.DefectLabelRoute = exports.DefectDefectExplorationItemRoute = exports.DefectComponentRoute = exports.DefectAttachmentRoute = exports.DefectRoute = exports.DataSetTestRoute = exports.DataSetRoute = exports.CouponSubscriptionsRoute = exports.CouponRoute = exports.ComponentDocTypeRoute = exports.ComponentDocTemplateRoute = exports.ComponentDocRoute = exports.ComponentComponentDocTypeRoute = exports.ComponentComponentDocTemplateRoute = exports.ComponentComponentDocRoute = exports.ComponentRoute = exports.CommentRoute = exports.ClientVirtualRoute = exports.ClientSignupOptionRoute = exports.ClientKeyRoute = void 0;
exports.PolicyPolicyRowsRoute = exports.PolicyRoute = exports.PlanWatchRoute = exports.PlanSuiteTestIncludeRoute = exports.PlanSuiteRoute = exports.PlanRunRoute = exports.PlanPurposeRoute = exports.PlanPlanSuiteTestIncludeRoute = exports.PlanLabelAssignedRoute = exports.PlanIntegrationProjectRoute = exports.PlanEnvironmentPlanRoute = exports.PlanCommentRoute = exports.PlanAttachmentRoute = exports.PlanRoute = exports.NotificationsRoute = exports.NativeDefectStatusLabelAssignedRoute = exports.NativeDefectStatusDefectStatusRoute = exports.NativeDefectStatusRoute = exports.NativeDefectResLabelAssignedRoute = exports.NativeDefectResDefectResRoute = exports.NativeDefectResRoute = exports.MilestoneTargetRoute = exports.MilestoneRunRoute = exports.MilestoneLabelAssignedRoute = exports.MilestoneExplorationRoute = exports.MilestoneCommentRoute = exports.MilestoneRoute = exports.LabelLabelProjectRoute = exports.LabelLabelAssignedRoute = exports.LabelAssignedRoute = exports.LabelRoute = exports.InvoicePaymentRoute = exports.InvoiceDisputeRoute = exports.InvoiceAttachmentRoute = exports.InvoiceRoute = exports.IntegrationUserRoute = exports.IntegrationTemplateRoute = exports.IntegrationStatusTypeNativeDefectStatusRoute = exports.IntegrationStatusTypeRoute = exports.IntegrationProjectRoute = exports.IntegrationCapabilityRoute = exports.IntegrationAppInstallIntegrationRoute = exports.IntegrationRoute = exports.FilterRoute = exports.ExplorationLabelAssignedRoute = exports.ExplorationItemLabelAssignedRoute = exports.ExplorationItemDefectExplorationItemRoute = exports.ExplorationItemRoute = exports.ExplorationExplorationItemRoute = exports.ExplorationEnvironmentExplorationRoute = void 0;
exports.PurposeLabelAssignedRoute = exports.PurposeRoute = exports.PullRequestPullRequestRunRoute = exports.PullRequestPullRequestRequirementRoute = exports.PullRequestCheckSuitePullRequestRoute = exports.PullRequestRoute = exports.ProjectWatchRoute = exports.ProjectTestRoute = exports.ProjectTaskRoute = exports.ProjectTargetRoute = exports.ProjectSuiteRoute = exports.ProjectStepRoute = exports.ProjectRunSuiteRoute = exports.ProjectRunResultStepRoute = exports.ProjectRunResultRoute = exports.ProjectRunRoute = exports.ProjectRequirementRoute = exports.ProjectProjectUserRoute = exports.ProjectProjectStatusRoute = exports.ProjectProjectPurposeRoute = exports.ProjectProjectCapabilityDefaultRoute = exports.ProjectPlanSuiteTestIncludeRoute = exports.ProjectPlanRoute = exports.ProjectMilestoneRoute = exports.ProjectLabelProjectRoute = exports.ProjectLabelAssignedRoute = exports.ProjectIntegrationRoute = exports.ProjectFilterRoute = exports.ProjectExplorationItemRoute = exports.ProjectExplorationRoute = exports.ProjectEnvironmentProjectRoute = exports.ProjectDocRoute = exports.ProjectDefectRoute = exports.ProjectDataSetRoute = exports.ProjectComponentDocRoute = exports.ProjectCommentRoute = exports.ProjectCodeRoute = exports.ProjectCaseTypeProjectRoute = exports.ProjectCasePriorityProjectRoute = exports.ProjectAttachmentRoute = exports.ProjectRoute = exports.ProductSubscriptionsRoute = exports.ProductMarketplaceRoute = exports.ProductRoute = exports.PriceSubscriptionsRoute = exports.PriceQuoteItemRoute = exports.PriceProductRoute = exports.PriceMarketplaceProductRoute = exports.PriceRoute = exports.PolicyRowsRoute = void 0;
exports.StepRunResultStepRoute = exports.StepLabelAssignedRoute = exports.StepCommentRoute = exports.StepRoute = exports.StatusTypeStatusRoute = exports.StatusTypeLabelAssignedRoute = exports.StatusTypeRoute = exports.StatusRunResultStepRoute = exports.StatusRunResultRoute = exports.StatusProjectStatusRoute = exports.StatusLabelAssignedRoute = exports.StatusExplorationItemRoute = exports.StatusExplorationRoute = exports.StatusRoute = exports.SignupOptionRoute = exports.SharedStepStepRoute = exports.SharedStepLabelAssignedRoute = exports.SharedStepRoute = exports.SharedPreconditionTestRoute = exports.SharedPreconditionRoute = exports.ShareShareUserRoute = exports.ShareRoute = exports.RunSuiteRoute = exports.RunRunSuiteRoute = exports.RunRunResultRoute = exports.RunResultStepLabelAssignedRoute = exports.RunResultStepCommentRoute = exports.RunResultStepAttachmentRoute = exports.RunResultStepRoute = exports.RunResultRunResultStepRoute = exports.RunResultLabelAssignedRoute = exports.RunResultDefectRoute = exports.RunResultCommentRoute = exports.RunResultAttachmentRoute = exports.RunResultRoute = exports.RunPullRequestRunRoute = exports.RunLabelAssignedRoute = exports.RunCommentRoute = exports.RunRoute = exports.ResourceValueEnvironmentResourceRoute = exports.ResourceValueRoute = exports.ResourceResourceValueRoute = exports.ResourceEnvironmentResourceRoute = exports.ResourceRoute = exports.RequirementRequirementTestRoute = exports.RequirementPullRequestRequirementRoute = exports.RequirementAttachmentRoute = exports.RequirementRoute = exports.PurposeProjectPurposeRoute = exports.PurposePlanRoute = void 0;
exports.UserSupportAgentRoute = exports.UserSupportRoute = exports.UserSubscriptionsRoute = exports.UserSubscriptionUserRoute = exports.UserShareUserRoute = exports.UserReportRoute = exports.UserQuoteRoute = exports.UserProjectUserRoute = exports.UserNotificationsRoute = exports.UserIntegrationRoute = exports.UserFeatureUserRoute = exports.UserExportRoute = exports.UserCheckListItemUserRoute = exports.UserBillingContactRoute = exports.UserAttachmentRoute = exports.UserAccessRoleRoute = exports.UserRoute = exports.TestWatchRoute = exports.TestSuiteRoute = exports.TestStepRoute = exports.TestRunResultRoute = exports.TestRequirementTestRoute = exports.TestQualityTypeTestQualityRoute = exports.TestQualityTypeLabelAssignedRoute = exports.TestQualityTypeRoute = exports.TestQualityTestRoute = exports.TestQualityLabelAssignedRoute = exports.TestQualityRoute = exports.TestPlanSuiteTestIncludeRoute = exports.TestLabelAssignedRoute = exports.TestCommentRoute = exports.TestAttachmentRoute = exports.TestRoute = exports.SuiteWatchRoute = exports.SuiteTestRoute = exports.SuiteRunResultRoute = exports.SuiteRequirementTestRoute = exports.SuitePlanSuiteTestIncludeRoute = exports.SuitePlanRoute = exports.SuiteLabelAssignedRoute = exports.SuiteCommentRoute = exports.SuiteRoute = exports.SubscriptionsSubscriptionUserRoute = exports.SubscriptionsQuoteItemRoute = exports.SubscriptionsBillingContactRoute = exports.SubscriptionsRoute = exports.SubscriptionUserRoute = exports.StripeProductProductRoute = exports.StripeProductPriceRoute = exports.StripeProductRoute = void 0;
exports.WebinarRoute = exports.WatchRoute = exports.VirtualRoute = exports.UserUserLogRoute = void 0;
const AccessRoleRoute = () => `/access_role`;
exports.AccessRoleRoute = AccessRoleRoute;
const AccessRolePolicyRoute = (access_role_id) => `/access_role/${access_role_id}/policy`;
exports.AccessRolePolicyRoute = AccessRolePolicyRoute;
const AccessRoleProjectRoute = (access_role_id) => `/access_role/${access_role_id}/project`;
exports.AccessRoleProjectRoute = AccessRoleProjectRoute;
const AccessRoleShareRoute = (access_role_id) => `/access_role/${access_role_id}/share`;
exports.AccessRoleShareRoute = AccessRoleShareRoute;
const AccessRoleUserRoute = (access_role_id) => `/access_role/${access_role_id}/user`;
exports.AccessRoleUserRoute = AccessRoleUserRoute;
const AppConfigRoute = () => `/app_config`;
exports.AppConfigRoute = AppConfigRoute;
const AppConfigAppInstallRoute = (app_config_id) => `/app_config/${app_config_id}/app_install`;
exports.AppConfigAppInstallRoute = AppConfigAppInstallRoute;
const AppInstallRoute = () => `/app_install`;
exports.AppInstallRoute = AppInstallRoute;
const AppInstallAppInstallIntegrationRoute = (app_install_id) => `/app_install/${app_install_id}/app_install_integration`;
exports.AppInstallAppInstallIntegrationRoute = AppInstallAppInstallIntegrationRoute;
const AppInstallAppInstallProjectRoute = (app_install_id) => `/app_install/${app_install_id}/app_install_project`;
exports.AppInstallAppInstallProjectRoute = AppInstallAppInstallProjectRoute;
const AppInstallProjectRoute = () => `/app_install_project`;
exports.AppInstallProjectRoute = AppInstallProjectRoute;
const AppInstallProjectBranchRoute = (app_install_project_id) => `/app_install_project/${app_install_project_id}/branch`;
exports.AppInstallProjectBranchRoute = AppInstallProjectBranchRoute;
const AppInstallProjectCheckRunRoute = (app_install_project_id) => `/app_install_project/${app_install_project_id}/check_run`;
exports.AppInstallProjectCheckRunRoute = AppInstallProjectCheckRunRoute;
const AppInstallProjectCheckSuiteRoute = (app_install_project_id) => `/app_install_project/${app_install_project_id}/check_suite`;
exports.AppInstallProjectCheckSuiteRoute = AppInstallProjectCheckSuiteRoute;
const AppInstallProjectIntegrationProjectRoute = (app_install_project_id) => `/app_install_project/${app_install_project_id}/integration_project`;
exports.AppInstallProjectIntegrationProjectRoute = AppInstallProjectIntegrationProjectRoute;
const AppInstallProjectPullRequestRoute = (app_install_project_id) => `/app_install_project/${app_install_project_id}/pull_request`;
exports.AppInstallProjectPullRequestRoute = AppInstallProjectPullRequestRoute;
const AppUserRoute = () => `/app_user`;
exports.AppUserRoute = AppUserRoute;
const AppUserAppInstallRoute = (app_user_id) => `/app_user/${app_user_id}/app_install`;
exports.AppUserAppInstallRoute = AppUserAppInstallRoute;
const AppUserBranchRoute = (app_user_id) => `/app_user/${app_user_id}/branch`;
exports.AppUserBranchRoute = AppUserBranchRoute;
const AppUserPullRequestRoute = (app_user_id) => `/app_user/${app_user_id}/pull_request`;
exports.AppUserPullRequestRoute = AppUserPullRequestRoute;
const AttachmentRoute = () => `/attachment`;
exports.AttachmentRoute = AttachmentRoute;
const BaseCapabilityRoute = () => `/base_capability`;
exports.BaseCapabilityRoute = BaseCapabilityRoute;
const BaseCapabilityBaseIntegrationRoute = (base_capability_id) => `/base_capability/${base_capability_id}/base_integration`;
exports.BaseCapabilityBaseIntegrationRoute = BaseCapabilityBaseIntegrationRoute;
const BaseCapabilityCapabilityRoute = (base_capability_id) => `/base_capability/${base_capability_id}/capability`;
exports.BaseCapabilityCapabilityRoute = BaseCapabilityCapabilityRoute;
const BaseIntegrationRoute = () => `/base_integration`;
exports.BaseIntegrationRoute = BaseIntegrationRoute;
const BaseIntegrationAppConfigRoute = (base_integration_id) => `/base_integration/${base_integration_id}/app_config`;
exports.BaseIntegrationAppConfigRoute = BaseIntegrationAppConfigRoute;
const BaseIntegrationBaseCapabilityRoute = (base_integration_id) => `/base_integration/${base_integration_id}/base_capability`;
exports.BaseIntegrationBaseCapabilityRoute = BaseIntegrationBaseCapabilityRoute;
const BaseIntegrationIntegrationRoute = (base_integration_id) => `/base_integration/${base_integration_id}/integration`;
exports.BaseIntegrationIntegrationRoute = BaseIntegrationIntegrationRoute;
const BillingContactRoute = () => `/billing_contact`;
exports.BillingContactRoute = BillingContactRoute;
const CapabilityRoute = () => `/capability`;
exports.CapabilityRoute = CapabilityRoute;
const CapabilityIntegrationRoute = (capability_id) => `/capability/${capability_id}/integration`;
exports.CapabilityIntegrationRoute = CapabilityIntegrationRoute;
const CapabilityProjectCapabilityDefaultRoute = (capability_id) => `/capability/${capability_id}/project_capability_default`;
exports.CapabilityProjectCapabilityDefaultRoute = CapabilityProjectCapabilityDefaultRoute;
const CasePriorityRoute = () => `/case_priority`;
exports.CasePriorityRoute = CasePriorityRoute;
const CasePriorityCasePriorityProjectRoute = (case_priority_id) => `/case_priority/${case_priority_id}/case_priority_project`;
exports.CasePriorityCasePriorityProjectRoute = CasePriorityCasePriorityProjectRoute;
const CasePriorityLabelAssignedRoute = (case_priority_id) => `/case_priority/${case_priority_id}/label_assigned`;
exports.CasePriorityLabelAssignedRoute = CasePriorityLabelAssignedRoute;
const CasePriorityTestRoute = (case_priority_id) => `/case_priority/${case_priority_id}/test`;
exports.CasePriorityTestRoute = CasePriorityTestRoute;
const CaseTypeRoute = () => `/case_type`;
exports.CaseTypeRoute = CaseTypeRoute;
const CaseTypeCaseTypeProjectRoute = (case_type_id) => `/case_type/${case_type_id}/case_type_project`;
exports.CaseTypeCaseTypeProjectRoute = CaseTypeCaseTypeProjectRoute;
const CaseTypeLabelAssignedRoute = (case_type_id) => `/case_type/${case_type_id}/label_assigned`;
exports.CaseTypeLabelAssignedRoute = CaseTypeLabelAssignedRoute;
const CaseTypeTestRoute = (case_type_id) => `/case_type/${case_type_id}/test`;
exports.CaseTypeTestRoute = CaseTypeTestRoute;
const CheckListRoute = () => `/check_list`;
exports.CheckListRoute = CheckListRoute;
const CheckListCheckListItemRoute = (check_list_id) => `/check_list/${check_list_id}/check_list_item`;
exports.CheckListCheckListItemRoute = CheckListCheckListItemRoute;
const CheckListItemRoute = () => `/check_list_item`;
exports.CheckListItemRoute = CheckListItemRoute;
const CheckListItemCheckListItemUserRoute = (check_list_item_id) => `/check_list_item/${check_list_item_id}/check_list_item_user`;
exports.CheckListItemCheckListItemUserRoute = CheckListItemCheckListItemUserRoute;
const CheckRunRoute = () => `/check_run`;
exports.CheckRunRoute = CheckRunRoute;
const CheckRunRunRoute = (check_run_id) => `/check_run/${check_run_id}/run`;
exports.CheckRunRunRoute = CheckRunRunRoute;
const CheckSuiteRoute = () => `/check_suite`;
exports.CheckSuiteRoute = CheckSuiteRoute;
const CheckSuiteCheckRunRoute = (check_suite_id) => `/check_suite/${check_suite_id}/check_run`;
exports.CheckSuiteCheckRunRoute = CheckSuiteCheckRunRoute;
const CheckSuiteCheckSuitePullRequestRoute = (check_suite_id) => `/check_suite/${check_suite_id}/check_suite_pull_request`;
exports.CheckSuiteCheckSuitePullRequestRoute = CheckSuiteCheckSuitePullRequestRoute;
const ClientRoute = () => `/client`;
exports.ClientRoute = ClientRoute;
const ClientKeyRoute = (client_id) => `/client/${client_id}/key`;
exports.ClientKeyRoute = ClientKeyRoute;
const ClientSignupOptionRoute = (client_id) => `/client/${client_id}/signup_option`;
exports.ClientSignupOptionRoute = ClientSignupOptionRoute;
const ClientVirtualRoute = (client_id) => `/client/${client_id}/virtual`;
exports.ClientVirtualRoute = ClientVirtualRoute;
const CommentRoute = () => `/comment`;
exports.CommentRoute = CommentRoute;
const ComponentRoute = () => `/component`;
exports.ComponentRoute = ComponentRoute;
const ComponentComponentDocRoute = (component_id) => `/component/${component_id}/component_doc`;
exports.ComponentComponentDocRoute = ComponentComponentDocRoute;
const ComponentComponentDocTemplateRoute = (component_id) => `/component/${component_id}/component_doc_template`;
exports.ComponentComponentDocTemplateRoute = ComponentComponentDocTemplateRoute;
const ComponentComponentDocTypeRoute = (component_id) => `/component/${component_id}/component_doc_type`;
exports.ComponentComponentDocTypeRoute = ComponentComponentDocTypeRoute;
const ComponentDocRoute = () => `/component_doc`;
exports.ComponentDocRoute = ComponentDocRoute;
const ComponentDocTemplateRoute = () => `/component_doc_template`;
exports.ComponentDocTemplateRoute = ComponentDocTemplateRoute;
const ComponentDocTypeRoute = () => `/component_doc_type`;
exports.ComponentDocTypeRoute = ComponentDocTypeRoute;
const CouponRoute = () => `/coupon`;
exports.CouponRoute = CouponRoute;
const CouponSubscriptionsRoute = (coupon_id) => `/coupon/${coupon_id}/subscriptions`;
exports.CouponSubscriptionsRoute = CouponSubscriptionsRoute;
const DataSetRoute = () => `/data_set`;
exports.DataSetRoute = DataSetRoute;
const DataSetTestRoute = (data_set_id) => `/data_set/${data_set_id}/test`;
exports.DataSetTestRoute = DataSetTestRoute;
const DefectRoute = () => `/defect`;
exports.DefectRoute = DefectRoute;
const DefectAttachmentRoute = (defect_id) => `/defect/${defect_id}/attachment`;
exports.DefectAttachmentRoute = DefectAttachmentRoute;
const DefectComponentRoute = () => `/defect_component`;
exports.DefectComponentRoute = DefectComponentRoute;
const DefectDefectExplorationItemRoute = (defect_id) => `/defect/${defect_id}/defect_exploration_item`;
exports.DefectDefectExplorationItemRoute = DefectDefectExplorationItemRoute;
const DefectLabelRoute = () => `/defect_label`;
exports.DefectLabelRoute = DefectLabelRoute;
const DefectPriorityRoute = () => `/defect_priority`;
exports.DefectPriorityRoute = DefectPriorityRoute;
const DefectResRoute = () => `/defect_res`;
exports.DefectResRoute = DefectResRoute;
const DefectResDefectRoute = (defect_res_id) => `/defect_res/${defect_res_id}/defect`;
exports.DefectResDefectRoute = DefectResDefectRoute;
const DefectResNativeDefectResRoute = (defect_res_id) => `/defect_res/${defect_res_id}/native_defect_res`;
exports.DefectResNativeDefectResRoute = DefectResNativeDefectResRoute;
const DefectResRequirementRoute = (defect_res_id) => `/defect_res/${defect_res_id}/requirement`;
exports.DefectResRequirementRoute = DefectResRequirementRoute;
const DefectRunResultRoute = (defect_id) => `/defect/${defect_id}/run_result`;
exports.DefectRunResultRoute = DefectRunResultRoute;
const DefectStatusRoute = () => `/defect_status`;
exports.DefectStatusRoute = DefectStatusRoute;
const DefectStatusDefectRoute = (defect_status_id) => `/defect_status/${defect_status_id}/defect`;
exports.DefectStatusDefectRoute = DefectStatusDefectRoute;
const DefectStatusNativeDefectStatusRoute = (defect_status_id) => `/defect_status/${defect_status_id}/native_defect_status`;
exports.DefectStatusNativeDefectStatusRoute = DefectStatusNativeDefectStatusRoute;
const DefectStatusRequirementRoute = (defect_status_id) => `/defect_status/${defect_status_id}/requirement`;
exports.DefectStatusRequirementRoute = DefectStatusRequirementRoute;
const DefectTypeRoute = () => `/defect_type`;
exports.DefectTypeRoute = DefectTypeRoute;
const DefectUserRoute = () => `/defect_user`;
exports.DefectUserRoute = DefectUserRoute;
const DocRoute = () => `/doc`;
exports.DocRoute = DocRoute;
const DocAttachmentRoute = (doc_id) => `/doc/${doc_id}/attachment`;
exports.DocAttachmentRoute = DocAttachmentRoute;
const DocComponentDocRoute = (doc_id) => `/doc/${doc_id}/component_doc`;
exports.DocComponentDocRoute = DocComponentDocRoute;
const DocTemplateRoute = () => `/doc_template`;
exports.DocTemplateRoute = DocTemplateRoute;
const DocTemplateComponentDocTemplateRoute = (doc_template_id) => `/doc_template/${doc_template_id}/component_doc_template`;
exports.DocTemplateComponentDocTemplateRoute = DocTemplateComponentDocTemplateRoute;
const DocTemplateDocRoute = (doc_template_id) => `/doc_template/${doc_template_id}/doc`;
exports.DocTemplateDocRoute = DocTemplateDocRoute;
const DocTypeRoute = () => `/doc_type`;
exports.DocTypeRoute = DocTypeRoute;
const DocTypeComponentDocTypeRoute = (doc_type_id) => `/doc_type/${doc_type_id}/component_doc_type`;
exports.DocTypeComponentDocTypeRoute = DocTypeComponentDocTypeRoute;
const DocTypeDocRoute = (doc_type_id) => `/doc_type/${doc_type_id}/doc`;
exports.DocTypeDocRoute = DocTypeDocRoute;
const DocTypeDocTemplateRoute = (doc_type_id) => `/doc_type/${doc_type_id}/doc_template`;
exports.DocTypeDocTemplateRoute = DocTypeDocTemplateRoute;
const EnvironmentRoute = () => `/environment`;
exports.EnvironmentRoute = EnvironmentRoute;
const EnvironmentEnvironmentExplorationRoute = (environment_id) => `/environment/${environment_id}/environment_exploration`;
exports.EnvironmentEnvironmentExplorationRoute = EnvironmentEnvironmentExplorationRoute;
const EnvironmentEnvironmentPlanRoute = (environment_id) => `/environment/${environment_id}/environment_plan`;
exports.EnvironmentEnvironmentPlanRoute = EnvironmentEnvironmentPlanRoute;
const EnvironmentEnvironmentProjectRoute = (environment_id) => `/environment/${environment_id}/environment_project`;
exports.EnvironmentEnvironmentProjectRoute = EnvironmentEnvironmentProjectRoute;
const EnvironmentEnvironmentResourceRoute = (environment_id) => `/environment/${environment_id}/environment_resource`;
exports.EnvironmentEnvironmentResourceRoute = EnvironmentEnvironmentResourceRoute;
const EnvironmentExplorationItemRoute = (environment_id) => `/environment/${environment_id}/exploration_item`;
exports.EnvironmentExplorationItemRoute = EnvironmentExplorationItemRoute;
const EnvironmentRunResultRoute = (environment_id) => `/environment/${environment_id}/run_result`;
exports.EnvironmentRunResultRoute = EnvironmentRunResultRoute;
const ExplorationRoute = () => `/exploration`;
exports.ExplorationRoute = ExplorationRoute;
const ExplorationEnvironmentExplorationRoute = (exploration_id) => `/exploration/${exploration_id}/environment_exploration`;
exports.ExplorationEnvironmentExplorationRoute = ExplorationEnvironmentExplorationRoute;
const ExplorationExplorationItemRoute = (exploration_id) => `/exploration/${exploration_id}/exploration_item`;
exports.ExplorationExplorationItemRoute = ExplorationExplorationItemRoute;
const ExplorationItemRoute = () => `/exploration_item`;
exports.ExplorationItemRoute = ExplorationItemRoute;
const ExplorationItemDefectExplorationItemRoute = (exploration_item_id) => `/exploration_item/${exploration_item_id}/defect_exploration_item`;
exports.ExplorationItemDefectExplorationItemRoute = ExplorationItemDefectExplorationItemRoute;
const ExplorationItemLabelAssignedRoute = (exploration_item_id) => `/exploration_item/${exploration_item_id}/label_assigned`;
exports.ExplorationItemLabelAssignedRoute = ExplorationItemLabelAssignedRoute;
const ExplorationLabelAssignedRoute = (exploration_id) => `/exploration/${exploration_id}/label_assigned`;
exports.ExplorationLabelAssignedRoute = ExplorationLabelAssignedRoute;
const FilterRoute = () => `/filter`;
exports.FilterRoute = FilterRoute;
const IntegrationRoute = () => `/integration`;
exports.IntegrationRoute = IntegrationRoute;
const IntegrationAppInstallIntegrationRoute = (integration_id) => `/integration/${integration_id}/app_install_integration`;
exports.IntegrationAppInstallIntegrationRoute = IntegrationAppInstallIntegrationRoute;
const IntegrationCapabilityRoute = (integration_id) => `/integration/${integration_id}/capability`;
exports.IntegrationCapabilityRoute = IntegrationCapabilityRoute;
const IntegrationProjectRoute = (integration_id) => `/integration/${integration_id}/project`;
exports.IntegrationProjectRoute = IntegrationProjectRoute;
const IntegrationStatusTypeRoute = () => `/integration_status_type`;
exports.IntegrationStatusTypeRoute = IntegrationStatusTypeRoute;
const IntegrationStatusTypeNativeDefectStatusRoute = (integration_status_type_id) => `/integration_status_type/${integration_status_type_id}/native_defect_status`;
exports.IntegrationStatusTypeNativeDefectStatusRoute = IntegrationStatusTypeNativeDefectStatusRoute;
const IntegrationTemplateRoute = () => `/integration_template`;
exports.IntegrationTemplateRoute = IntegrationTemplateRoute;
const IntegrationUserRoute = (integration_id) => `/integration/${integration_id}/user`;
exports.IntegrationUserRoute = IntegrationUserRoute;
const InvoiceRoute = () => `/invoice`;
exports.InvoiceRoute = InvoiceRoute;
const InvoiceAttachmentRoute = (invoice_id) => `/invoice/${invoice_id}/attachment`;
exports.InvoiceAttachmentRoute = InvoiceAttachmentRoute;
const InvoiceDisputeRoute = (invoice_id) => `/invoice/${invoice_id}/dispute`;
exports.InvoiceDisputeRoute = InvoiceDisputeRoute;
const InvoicePaymentRoute = (invoice_id) => `/invoice/${invoice_id}/payment`;
exports.InvoicePaymentRoute = InvoicePaymentRoute;
const LabelRoute = () => `/label`;
exports.LabelRoute = LabelRoute;
const LabelAssignedRoute = () => `/label_assigned`;
exports.LabelAssignedRoute = LabelAssignedRoute;
const LabelLabelAssignedRoute = (label_id) => `/label/${label_id}/label_assigned`;
exports.LabelLabelAssignedRoute = LabelLabelAssignedRoute;
const LabelLabelProjectRoute = (label_id) => `/label/${label_id}/label_project`;
exports.LabelLabelProjectRoute = LabelLabelProjectRoute;
const MilestoneRoute = () => `/milestone`;
exports.MilestoneRoute = MilestoneRoute;
const MilestoneCommentRoute = (milestone_id) => `/milestone/${milestone_id}/comment`;
exports.MilestoneCommentRoute = MilestoneCommentRoute;
const MilestoneExplorationRoute = (milestone_id) => `/milestone/${milestone_id}/exploration`;
exports.MilestoneExplorationRoute = MilestoneExplorationRoute;
const MilestoneLabelAssignedRoute = (milestone_id) => `/milestone/${milestone_id}/label_assigned`;
exports.MilestoneLabelAssignedRoute = MilestoneLabelAssignedRoute;
const MilestoneRunRoute = (milestone_id) => `/milestone/${milestone_id}/run`;
exports.MilestoneRunRoute = MilestoneRunRoute;
const MilestoneTargetRoute = (milestone_id) => `/milestone/${milestone_id}/target`;
exports.MilestoneTargetRoute = MilestoneTargetRoute;
const NativeDefectResRoute = () => `/native_defect_res`;
exports.NativeDefectResRoute = NativeDefectResRoute;
const NativeDefectResDefectResRoute = (native_defect_res_id) => `/native_defect_res/${native_defect_res_id}/defect_res`;
exports.NativeDefectResDefectResRoute = NativeDefectResDefectResRoute;
const NativeDefectResLabelAssignedRoute = (native_defect_res_id) => `/native_defect_res/${native_defect_res_id}/label_assigned`;
exports.NativeDefectResLabelAssignedRoute = NativeDefectResLabelAssignedRoute;
const NativeDefectStatusRoute = () => `/native_defect_status`;
exports.NativeDefectStatusRoute = NativeDefectStatusRoute;
const NativeDefectStatusDefectStatusRoute = (native_defect_status_id) => `/native_defect_status/${native_defect_status_id}/defect_status`;
exports.NativeDefectStatusDefectStatusRoute = NativeDefectStatusDefectStatusRoute;
const NativeDefectStatusLabelAssignedRoute = (native_defect_status_id) => `/native_defect_status/${native_defect_status_id}/label_assigned`;
exports.NativeDefectStatusLabelAssignedRoute = NativeDefectStatusLabelAssignedRoute;
const NotificationsRoute = () => `/notifications`;
exports.NotificationsRoute = NotificationsRoute;
const PlanRoute = () => `/plan`;
exports.PlanRoute = PlanRoute;
const PlanAttachmentRoute = (plan_id) => `/plan/${plan_id}/attachment`;
exports.PlanAttachmentRoute = PlanAttachmentRoute;
const PlanCommentRoute = (plan_id) => `/plan/${plan_id}/comment`;
exports.PlanCommentRoute = PlanCommentRoute;
const PlanEnvironmentPlanRoute = (plan_id) => `/plan/${plan_id}/environment_plan`;
exports.PlanEnvironmentPlanRoute = PlanEnvironmentPlanRoute;
const PlanIntegrationProjectRoute = (plan_id) => `/plan/${plan_id}/integration_project`;
exports.PlanIntegrationProjectRoute = PlanIntegrationProjectRoute;
const PlanLabelAssignedRoute = (plan_id) => `/plan/${plan_id}/label_assigned`;
exports.PlanLabelAssignedRoute = PlanLabelAssignedRoute;
const PlanPlanSuiteTestIncludeRoute = (plan_id) => `/plan/${plan_id}/plan_suite_test_include`;
exports.PlanPlanSuiteTestIncludeRoute = PlanPlanSuiteTestIncludeRoute;
const PlanPurposeRoute = (plan_id) => `/plan/${plan_id}/purpose`;
exports.PlanPurposeRoute = PlanPurposeRoute;
const PlanRunRoute = (plan_id) => `/plan/${plan_id}/run`;
exports.PlanRunRoute = PlanRunRoute;
const PlanSuiteRoute = (plan_id) => `/plan/${plan_id}/suite`;
exports.PlanSuiteRoute = PlanSuiteRoute;
const PlanSuiteTestIncludeRoute = () => `/plan_suite_test_include`;
exports.PlanSuiteTestIncludeRoute = PlanSuiteTestIncludeRoute;
const PlanWatchRoute = (plan_id) => `/plan/${plan_id}/watch`;
exports.PlanWatchRoute = PlanWatchRoute;
const PolicyRoute = () => `/policy`;
exports.PolicyRoute = PolicyRoute;
const PolicyPolicyRowsRoute = (policy_id) => `/policy/${policy_id}/policy_rows`;
exports.PolicyPolicyRowsRoute = PolicyPolicyRowsRoute;
const PolicyRowsRoute = () => `/policy_rows`;
exports.PolicyRowsRoute = PolicyRowsRoute;
const PriceRoute = () => `/price`;
exports.PriceRoute = PriceRoute;
const PriceMarketplaceProductRoute = (price_id) => `/price/${price_id}/marketplace_product`;
exports.PriceMarketplaceProductRoute = PriceMarketplaceProductRoute;
const PriceProductRoute = (price_id) => `/price/${price_id}/product`;
exports.PriceProductRoute = PriceProductRoute;
const PriceQuoteItemRoute = (price_id) => `/price/${price_id}/quote_item`;
exports.PriceQuoteItemRoute = PriceQuoteItemRoute;
const PriceSubscriptionsRoute = (price_id) => `/price/${price_id}/subscriptions`;
exports.PriceSubscriptionsRoute = PriceSubscriptionsRoute;
const ProductRoute = () => `/product`;
exports.ProductRoute = ProductRoute;
const ProductMarketplaceRoute = (product_id) => `/product/${product_id}/marketplace`;
exports.ProductMarketplaceRoute = ProductMarketplaceRoute;
const ProductSubscriptionsRoute = (product_id) => `/product/${product_id}/subscriptions`;
exports.ProductSubscriptionsRoute = ProductSubscriptionsRoute;
const ProjectRoute = () => `/project`;
exports.ProjectRoute = ProjectRoute;
const ProjectAttachmentRoute = (project_id) => `/project/${project_id}/attachment`;
exports.ProjectAttachmentRoute = ProjectAttachmentRoute;
const ProjectCasePriorityProjectRoute = (project_id) => `/project/${project_id}/case_priority_project`;
exports.ProjectCasePriorityProjectRoute = ProjectCasePriorityProjectRoute;
const ProjectCaseTypeProjectRoute = (project_id) => `/project/${project_id}/case_type_project`;
exports.ProjectCaseTypeProjectRoute = ProjectCaseTypeProjectRoute;
const ProjectCodeRoute = (project_id) => `/project/${project_id}/code`;
exports.ProjectCodeRoute = ProjectCodeRoute;
const ProjectCommentRoute = (project_id) => `/project/${project_id}/comment`;
exports.ProjectCommentRoute = ProjectCommentRoute;
const ProjectComponentDocRoute = (project_id) => `/project/${project_id}/component_doc`;
exports.ProjectComponentDocRoute = ProjectComponentDocRoute;
const ProjectDataSetRoute = (project_id) => `/project/${project_id}/data_set`;
exports.ProjectDataSetRoute = ProjectDataSetRoute;
const ProjectDefectRoute = (project_id) => `/project/${project_id}/defect`;
exports.ProjectDefectRoute = ProjectDefectRoute;
const ProjectDocRoute = (project_id) => `/project/${project_id}/doc`;
exports.ProjectDocRoute = ProjectDocRoute;
const ProjectEnvironmentProjectRoute = (project_id) => `/project/${project_id}/environment_project`;
exports.ProjectEnvironmentProjectRoute = ProjectEnvironmentProjectRoute;
const ProjectExplorationRoute = (project_id) => `/project/${project_id}/exploration`;
exports.ProjectExplorationRoute = ProjectExplorationRoute;
const ProjectExplorationItemRoute = (project_id) => `/project/${project_id}/exploration_item`;
exports.ProjectExplorationItemRoute = ProjectExplorationItemRoute;
const ProjectFilterRoute = (project_id) => `/project/${project_id}/filter`;
exports.ProjectFilterRoute = ProjectFilterRoute;
const ProjectIntegrationRoute = (project_id) => `/project/${project_id}/integration`;
exports.ProjectIntegrationRoute = ProjectIntegrationRoute;
const ProjectLabelAssignedRoute = (project_id) => `/project/${project_id}/label_assigned`;
exports.ProjectLabelAssignedRoute = ProjectLabelAssignedRoute;
const ProjectLabelProjectRoute = (project_id) => `/project/${project_id}/label_project`;
exports.ProjectLabelProjectRoute = ProjectLabelProjectRoute;
const ProjectMilestoneRoute = (project_id) => `/project/${project_id}/milestone`;
exports.ProjectMilestoneRoute = ProjectMilestoneRoute;
const ProjectPlanRoute = (project_id) => `/project/${project_id}/plan`;
exports.ProjectPlanRoute = ProjectPlanRoute;
const ProjectPlanSuiteTestIncludeRoute = (project_id) => `/project/${project_id}/plan_suite_test_include`;
exports.ProjectPlanSuiteTestIncludeRoute = ProjectPlanSuiteTestIncludeRoute;
const ProjectProjectCapabilityDefaultRoute = (project_id) => `/project/${project_id}/project_capability_default`;
exports.ProjectProjectCapabilityDefaultRoute = ProjectProjectCapabilityDefaultRoute;
const ProjectProjectPurposeRoute = (project_id) => `/project/${project_id}/project_purpose`;
exports.ProjectProjectPurposeRoute = ProjectProjectPurposeRoute;
const ProjectProjectStatusRoute = (project_id) => `/project/${project_id}/project_status`;
exports.ProjectProjectStatusRoute = ProjectProjectStatusRoute;
const ProjectProjectUserRoute = (project_id) => `/project/${project_id}/project_user`;
exports.ProjectProjectUserRoute = ProjectProjectUserRoute;
const ProjectRequirementRoute = (project_id) => `/project/${project_id}/requirement`;
exports.ProjectRequirementRoute = ProjectRequirementRoute;
const ProjectRunRoute = (project_id) => `/project/${project_id}/run`;
exports.ProjectRunRoute = ProjectRunRoute;
const ProjectRunResultRoute = (project_id) => `/project/${project_id}/run_result`;
exports.ProjectRunResultRoute = ProjectRunResultRoute;
const ProjectRunResultStepRoute = (project_id) => `/project/${project_id}/run_result_step`;
exports.ProjectRunResultStepRoute = ProjectRunResultStepRoute;
const ProjectRunSuiteRoute = (project_id) => `/project/${project_id}/run_suite`;
exports.ProjectRunSuiteRoute = ProjectRunSuiteRoute;
const ProjectStepRoute = (project_id) => `/project/${project_id}/step`;
exports.ProjectStepRoute = ProjectStepRoute;
const ProjectSuiteRoute = (project_id) => `/project/${project_id}/suite`;
exports.ProjectSuiteRoute = ProjectSuiteRoute;
const ProjectTargetRoute = (project_id) => `/project/${project_id}/target`;
exports.ProjectTargetRoute = ProjectTargetRoute;
const ProjectTaskRoute = (project_id) => `/project/${project_id}/task`;
exports.ProjectTaskRoute = ProjectTaskRoute;
const ProjectTestRoute = (project_id) => `/project/${project_id}/test`;
exports.ProjectTestRoute = ProjectTestRoute;
const ProjectWatchRoute = (project_id) => `/project/${project_id}/watch`;
exports.ProjectWatchRoute = ProjectWatchRoute;
const PullRequestRoute = () => `/pull_request`;
exports.PullRequestRoute = PullRequestRoute;
const PullRequestCheckSuitePullRequestRoute = (pull_request_id) => `/pull_request/${pull_request_id}/check_suite_pull_request`;
exports.PullRequestCheckSuitePullRequestRoute = PullRequestCheckSuitePullRequestRoute;
const PullRequestPullRequestRequirementRoute = (pull_request_id) => `/pull_request/${pull_request_id}/pull_request_requirement`;
exports.PullRequestPullRequestRequirementRoute = PullRequestPullRequestRequirementRoute;
const PullRequestPullRequestRunRoute = (pull_request_id) => `/pull_request/${pull_request_id}/pull_request_run`;
exports.PullRequestPullRequestRunRoute = PullRequestPullRequestRunRoute;
const PurposeRoute = () => `/purpose`;
exports.PurposeRoute = PurposeRoute;
const PurposeLabelAssignedRoute = (purpose_id) => `/purpose/${purpose_id}/label_assigned`;
exports.PurposeLabelAssignedRoute = PurposeLabelAssignedRoute;
const PurposePlanRoute = (purpose_id) => `/purpose/${purpose_id}/plan`;
exports.PurposePlanRoute = PurposePlanRoute;
const PurposeProjectPurposeRoute = (purpose_id) => `/purpose/${purpose_id}/project_purpose`;
exports.PurposeProjectPurposeRoute = PurposeProjectPurposeRoute;
const RequirementRoute = () => `/requirement`;
exports.RequirementRoute = RequirementRoute;
const RequirementAttachmentRoute = (requirement_id) => `/requirement/${requirement_id}/attachment`;
exports.RequirementAttachmentRoute = RequirementAttachmentRoute;
const RequirementPullRequestRequirementRoute = (requirement_id) => `/requirement/${requirement_id}/pull_request_requirement`;
exports.RequirementPullRequestRequirementRoute = RequirementPullRequestRequirementRoute;
const RequirementRequirementTestRoute = (requirement_id) => `/requirement/${requirement_id}/requirement_test`;
exports.RequirementRequirementTestRoute = RequirementRequirementTestRoute;
const ResourceRoute = () => `/resource`;
exports.ResourceRoute = ResourceRoute;
const ResourceEnvironmentResourceRoute = (resource_id) => `/resource/${resource_id}/environment_resource`;
exports.ResourceEnvironmentResourceRoute = ResourceEnvironmentResourceRoute;
const ResourceResourceValueRoute = (resource_id) => `/resource/${resource_id}/resource_value`;
exports.ResourceResourceValueRoute = ResourceResourceValueRoute;
const ResourceValueRoute = () => `/resource_value`;
exports.ResourceValueRoute = ResourceValueRoute;
const ResourceValueEnvironmentResourceRoute = (resource_value_id) => `/resource_value/${resource_value_id}/environment_resource`;
exports.ResourceValueEnvironmentResourceRoute = ResourceValueEnvironmentResourceRoute;
const RunRoute = () => `/run`;
exports.RunRoute = RunRoute;
const RunCommentRoute = (run_id) => `/run/${run_id}/comment`;
exports.RunCommentRoute = RunCommentRoute;
const RunLabelAssignedRoute = (run_id) => `/run/${run_id}/label_assigned`;
exports.RunLabelAssignedRoute = RunLabelAssignedRoute;
const RunPullRequestRunRoute = (run_id) => `/run/${run_id}/pull_request_run`;
exports.RunPullRequestRunRoute = RunPullRequestRunRoute;
const RunResultRoute = () => `/run_result`;
exports.RunResultRoute = RunResultRoute;
const RunResultAttachmentRoute = (run_result_id) => `/run_result/${run_result_id}/attachment`;
exports.RunResultAttachmentRoute = RunResultAttachmentRoute;
const RunResultCommentRoute = (run_result_id) => `/run_result/${run_result_id}/comment`;
exports.RunResultCommentRoute = RunResultCommentRoute;
const RunResultDefectRoute = (run_result_id) => `/run_result/${run_result_id}/defect`;
exports.RunResultDefectRoute = RunResultDefectRoute;
const RunResultLabelAssignedRoute = (run_result_id) => `/run_result/${run_result_id}/label_assigned`;
exports.RunResultLabelAssignedRoute = RunResultLabelAssignedRoute;
const RunResultRunResultStepRoute = (run_result_id) => `/run_result/${run_result_id}/run_result_step`;
exports.RunResultRunResultStepRoute = RunResultRunResultStepRoute;
const RunResultStepRoute = () => `/run_result_step`;
exports.RunResultStepRoute = RunResultStepRoute;
const RunResultStepAttachmentRoute = (run_result_step_id) => `/run_result_step/${run_result_step_id}/attachment`;
exports.RunResultStepAttachmentRoute = RunResultStepAttachmentRoute;
const RunResultStepCommentRoute = (run_result_step_id) => `/run_result_step/${run_result_step_id}/comment`;
exports.RunResultStepCommentRoute = RunResultStepCommentRoute;
const RunResultStepLabelAssignedRoute = (run_result_step_id) => `/run_result_step/${run_result_step_id}/label_assigned`;
exports.RunResultStepLabelAssignedRoute = RunResultStepLabelAssignedRoute;
const RunRunResultRoute = (run_id) => `/run/${run_id}/run_result`;
exports.RunRunResultRoute = RunRunResultRoute;
const RunRunSuiteRoute = (run_id) => `/run/${run_id}/run_suite`;
exports.RunRunSuiteRoute = RunRunSuiteRoute;
const RunSuiteRoute = () => `/run_suite`;
exports.RunSuiteRoute = RunSuiteRoute;
const ShareRoute = () => `/share`;
exports.ShareRoute = ShareRoute;
const ShareShareUserRoute = (share_id) => `/share/${share_id}/share_user`;
exports.ShareShareUserRoute = ShareShareUserRoute;
const SharedPreconditionRoute = () => `/shared_precondition`;
exports.SharedPreconditionRoute = SharedPreconditionRoute;
const SharedPreconditionTestRoute = (shared_precondition_id) => `/shared_precondition/${shared_precondition_id}/test`;
exports.SharedPreconditionTestRoute = SharedPreconditionTestRoute;
const SharedStepRoute = () => `/shared_step`;
exports.SharedStepRoute = SharedStepRoute;
const SharedStepLabelAssignedRoute = (shared_step_id) => `/shared_step/${shared_step_id}/label_assigned`;
exports.SharedStepLabelAssignedRoute = SharedStepLabelAssignedRoute;
const SharedStepStepRoute = (shared_step_id) => `/shared_step/${shared_step_id}/step`;
exports.SharedStepStepRoute = SharedStepStepRoute;
const SignupOptionRoute = () => `/signup_option`;
exports.SignupOptionRoute = SignupOptionRoute;
const StatusRoute = () => `/status`;
exports.StatusRoute = StatusRoute;
const StatusExplorationRoute = (status_id) => `/status/${status_id}/exploration`;
exports.StatusExplorationRoute = StatusExplorationRoute;
const StatusExplorationItemRoute = (status_id) => `/status/${status_id}/exploration_item`;
exports.StatusExplorationItemRoute = StatusExplorationItemRoute;
const StatusLabelAssignedRoute = (status_id) => `/status/${status_id}/label_assigned`;
exports.StatusLabelAssignedRoute = StatusLabelAssignedRoute;
const StatusProjectStatusRoute = (status_id) => `/status/${status_id}/project_status`;
exports.StatusProjectStatusRoute = StatusProjectStatusRoute;
const StatusRunResultRoute = (status_id) => `/status/${status_id}/run_result`;
exports.StatusRunResultRoute = StatusRunResultRoute;
const StatusRunResultStepRoute = (status_id) => `/status/${status_id}/run_result_step`;
exports.StatusRunResultStepRoute = StatusRunResultStepRoute;
const StatusTypeRoute = () => `/status_type`;
exports.StatusTypeRoute = StatusTypeRoute;
const StatusTypeLabelAssignedRoute = (status_type_id) => `/status_type/${status_type_id}/label_assigned`;
exports.StatusTypeLabelAssignedRoute = StatusTypeLabelAssignedRoute;
const StatusTypeStatusRoute = (status_type_id) => `/status_type/${status_type_id}/status`;
exports.StatusTypeStatusRoute = StatusTypeStatusRoute;
const StepRoute = () => `/step`;
exports.StepRoute = StepRoute;
const StepCommentRoute = (step_id) => `/step/${step_id}/comment`;
exports.StepCommentRoute = StepCommentRoute;
const StepLabelAssignedRoute = (step_id) => `/step/${step_id}/label_assigned`;
exports.StepLabelAssignedRoute = StepLabelAssignedRoute;
const StepRunResultStepRoute = (step_id) => `/step/${step_id}/run_result_step`;
exports.StepRunResultStepRoute = StepRunResultStepRoute;
const StripeProductRoute = () => `/stripe_product`;
exports.StripeProductRoute = StripeProductRoute;
const StripeProductPriceRoute = (stripe_product_id) => `/stripe_product/${stripe_product_id}/price`;
exports.StripeProductPriceRoute = StripeProductPriceRoute;
const StripeProductProductRoute = (stripe_product_id) => `/stripe_product/${stripe_product_id}/product`;
exports.StripeProductProductRoute = StripeProductProductRoute;
const SubscriptionUserRoute = () => `/subscription_user`;
exports.SubscriptionUserRoute = SubscriptionUserRoute;
const SubscriptionsRoute = () => `/subscriptions`;
exports.SubscriptionsRoute = SubscriptionsRoute;
const SubscriptionsBillingContactRoute = (subscriptions_id) => `/subscriptions/${subscriptions_id}/billing_contact`;
exports.SubscriptionsBillingContactRoute = SubscriptionsBillingContactRoute;
const SubscriptionsQuoteItemRoute = (subscriptions_id) => `/subscriptions/${subscriptions_id}/quote_item`;
exports.SubscriptionsQuoteItemRoute = SubscriptionsQuoteItemRoute;
const SubscriptionsSubscriptionUserRoute = (subscriptions_id) => `/subscriptions/${subscriptions_id}/subscription_user`;
exports.SubscriptionsSubscriptionUserRoute = SubscriptionsSubscriptionUserRoute;
const SuiteRoute = () => `/suite`;
exports.SuiteRoute = SuiteRoute;
const SuiteCommentRoute = (suite_id) => `/suite/${suite_id}/comment`;
exports.SuiteCommentRoute = SuiteCommentRoute;
const SuiteLabelAssignedRoute = (suite_id) => `/suite/${suite_id}/label_assigned`;
exports.SuiteLabelAssignedRoute = SuiteLabelAssignedRoute;
const SuitePlanRoute = (suite_id) => `/suite/${suite_id}/plan`;
exports.SuitePlanRoute = SuitePlanRoute;
const SuitePlanSuiteTestIncludeRoute = (suite_id) => `/suite/${suite_id}/plan_suite_test_include`;
exports.SuitePlanSuiteTestIncludeRoute = SuitePlanSuiteTestIncludeRoute;
const SuiteRequirementTestRoute = (suite_id) => `/suite/${suite_id}/requirement_test`;
exports.SuiteRequirementTestRoute = SuiteRequirementTestRoute;
const SuiteRunResultRoute = (suite_id) => `/suite/${suite_id}/run_result`;
exports.SuiteRunResultRoute = SuiteRunResultRoute;
const SuiteTestRoute = (suite_id) => `/suite/${suite_id}/test`;
exports.SuiteTestRoute = SuiteTestRoute;
const SuiteWatchRoute = (suite_id) => `/suite/${suite_id}/watch`;
exports.SuiteWatchRoute = SuiteWatchRoute;
const TestRoute = () => `/test`;
exports.TestRoute = TestRoute;
const TestAttachmentRoute = (test_id) => `/test/${test_id}/attachment`;
exports.TestAttachmentRoute = TestAttachmentRoute;
const TestCommentRoute = (test_id) => `/test/${test_id}/comment`;
exports.TestCommentRoute = TestCommentRoute;
const TestLabelAssignedRoute = (test_id) => `/test/${test_id}/label_assigned`;
exports.TestLabelAssignedRoute = TestLabelAssignedRoute;
const TestPlanSuiteTestIncludeRoute = (test_id) => `/test/${test_id}/plan_suite_test_include`;
exports.TestPlanSuiteTestIncludeRoute = TestPlanSuiteTestIncludeRoute;
const TestQualityRoute = () => `/test_quality`;
exports.TestQualityRoute = TestQualityRoute;
const TestQualityLabelAssignedRoute = (test_quality_id) => `/test_quality/${test_quality_id}/label_assigned`;
exports.TestQualityLabelAssignedRoute = TestQualityLabelAssignedRoute;
const TestQualityTestRoute = (test_quality_id) => `/test_quality/${test_quality_id}/test`;
exports.TestQualityTestRoute = TestQualityTestRoute;
const TestQualityTypeRoute = () => `/test_quality_type`;
exports.TestQualityTypeRoute = TestQualityTypeRoute;
const TestQualityTypeLabelAssignedRoute = (test_quality_type_id) => `/test_quality_type/${test_quality_type_id}/label_assigned`;
exports.TestQualityTypeLabelAssignedRoute = TestQualityTypeLabelAssignedRoute;
const TestQualityTypeTestQualityRoute = (test_quality_type_id) => `/test_quality_type/${test_quality_type_id}/test_quality`;
exports.TestQualityTypeTestQualityRoute = TestQualityTypeTestQualityRoute;
const TestRequirementTestRoute = (test_id) => `/test/${test_id}/requirement_test`;
exports.TestRequirementTestRoute = TestRequirementTestRoute;
const TestRunResultRoute = (test_id) => `/test/${test_id}/run_result`;
exports.TestRunResultRoute = TestRunResultRoute;
const TestStepRoute = (test_id) => `/test/${test_id}/step`;
exports.TestStepRoute = TestStepRoute;
const TestSuiteRoute = (test_id) => `/test/${test_id}/suite`;
exports.TestSuiteRoute = TestSuiteRoute;
const TestWatchRoute = (test_id) => `/test/${test_id}/watch`;
exports.TestWatchRoute = TestWatchRoute;
const UserRoute = () => `/user`;
exports.UserRoute = UserRoute;
const UserAccessRoleRoute = (user_id) => `/user/${user_id}/access_role`;
exports.UserAccessRoleRoute = UserAccessRoleRoute;
const UserAttachmentRoute = (user_id) => `/user/${user_id}/attachment`;
exports.UserAttachmentRoute = UserAttachmentRoute;
const UserBillingContactRoute = (user_id) => `/user/${user_id}/billing_contact`;
exports.UserBillingContactRoute = UserBillingContactRoute;
const UserCheckListItemUserRoute = (user_id) => `/user/${user_id}/check_list_item_user`;
exports.UserCheckListItemUserRoute = UserCheckListItemUserRoute;
const UserExportRoute = (user_id) => `/user/${user_id}/export`;
exports.UserExportRoute = UserExportRoute;
const UserFeatureUserRoute = (user_id) => `/user/${user_id}/feature_user`;
exports.UserFeatureUserRoute = UserFeatureUserRoute;
const UserIntegrationRoute = (user_id) => `/user/${user_id}/integration`;
exports.UserIntegrationRoute = UserIntegrationRoute;
const UserNotificationsRoute = (user_id) => `/user/${user_id}/notifications`;
exports.UserNotificationsRoute = UserNotificationsRoute;
const UserProjectUserRoute = (user_id) => `/user/${user_id}/project_user`;
exports.UserProjectUserRoute = UserProjectUserRoute;
const UserQuoteRoute = (user_id) => `/user/${user_id}/quote`;
exports.UserQuoteRoute = UserQuoteRoute;
const UserReportRoute = (user_id) => `/user/${user_id}/report`;
exports.UserReportRoute = UserReportRoute;
const UserShareUserRoute = (user_id) => `/user/${user_id}/share_user`;
exports.UserShareUserRoute = UserShareUserRoute;
const UserSubscriptionUserRoute = (user_id) => `/user/${user_id}/subscription_user`;
exports.UserSubscriptionUserRoute = UserSubscriptionUserRoute;
const UserSubscriptionsRoute = (user_id) => `/user/${user_id}/subscriptions`;
exports.UserSubscriptionsRoute = UserSubscriptionsRoute;
const UserSupportRoute = (user_id) => `/user/${user_id}/support`;
exports.UserSupportRoute = UserSupportRoute;
const UserSupportAgentRoute = (user_id) => `/user/${user_id}/support_agent`;
exports.UserSupportAgentRoute = UserSupportAgentRoute;
const UserUserLogRoute = (user_id) => `/user/${user_id}/user_log`;
exports.UserUserLogRoute = UserUserLogRoute;
const VirtualRoute = () => `/virtual`;
exports.VirtualRoute = VirtualRoute;
const WatchRoute = () => `/watch`;
exports.WatchRoute = WatchRoute;
const WebinarRoute = () => `/webinar`;
exports.WebinarRoute = WebinarRoute;
