// @testquality/sdk/dist/services/role/accessRoleProject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.accessRoleProjectGetMany = void 0;
const ClientSdk_1 = require("../../ClientSdk");
const actions_1 = require("../../gen/actions");
const accessRoleProjectGetMany = (queryParams) => {
    var _a, _b;
    const config = {
        method: 'get',
        url: (_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.url) !== null && _a !== void 0 ? _a : '/access_role/project',
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, actions_1.getResponse)((_b = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _b !== void 0 ? _b : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.accessRoleProjectGetMany = accessRoleProjectGetMany;
