// @testquality/sdk/dist/services/role/policyGroupService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.policyGroupUpdateOne = exports.policyGroupGetMany = exports.PolicyGroupActions = void 0;
const actions_1 = require("../../gen/actions");
const ClientSdk_1 = require("../../ClientSdk");
var PolicyGroupActions;
(function (PolicyGroupActions) {
    PolicyGroupActions["view"] = "view";
    PolicyGroupActions["edit"] = "edit";
})(PolicyGroupActions || (exports.PolicyGroupActions = PolicyGroupActions = {}));
const policyGroupGetMany = (queryParams) => {
    var _a;
    const config = {
        method: 'get',
        url: '/policy_group',
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        cancelToken: queryParams === null || queryParams === void 0 ? void 0 : queryParams.cancelToken,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, actions_1.getResponse)((_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _a !== void 0 ? _a : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.policyGroupGetMany = policyGroupGetMany;
const policyGroupUpdateOne = (id, data, queryParams) => {
    var _a;
    const config = {
        method: 'put',
        url: `/policy_group/${id}`,
        params: queryParams === null || queryParams === void 0 ? void 0 : queryParams.params,
        data,
        headers: queryParams === null || queryParams === void 0 ? void 0 : queryParams.headers,
    };
    return (queryParams === null || queryParams === void 0 ? void 0 : queryParams.batch)
        ? queryParams.batch.addBatch(config)
        : (0, actions_1.getResponse)((_a = queryParams === null || queryParams === void 0 ? void 0 : queryParams.api) !== null && _a !== void 0 ? _a : ClientSdk_1._client === null || ClientSdk_1._client === void 0 ? void 0 : ClientSdk_1._client.api, config);
};
exports.policyGroupUpdateOne = policyGroupUpdateOne;
